File: orion/backend/lattigo/evaluator.go
package main

import (
	"C"

	"github.com/baahl-nyu/lattigo/v6/core/rlwe"
	"github.com/baahl-nyu/lattigo/v6/schemes/ckks"
)

var liveRotKeys = make(map[uint64]*rlwe.GaloisKey)
var savedRotKeys = []uint64{}

//export NewEvaluator
func NewEvaluator() {
	scheme.Evaluator = ckks.NewEvaluator(
		*scheme.Params, rlwe.NewMemEvaluationKeySet(scheme.RelinKey))

	// After declaring the evaluator, we'll also just generate and
	// store in memory all power of two rotation keys. This will ensure
	// all keys needed for the rotations and summations in the hyrid
	// method remain alive.
	AddPo2RotationKeys()
}

func AddPo2RotationKeys() {
	maxSlots := scheme.Params.MaxSlots()
	// Generate all positive power-of-two rotation keys, including maxSlots itself
	for i := 1; i <= maxSlots; i *= 2 {
		AddRotationKey(C.int(i))
	}
}

//export AddRotationKey
func AddRotationKey(rotation C.int) {
	galEl := scheme.Params.GaloisElement(int(rotation))

	// Check if key exists in scheme.EvalKeys.GaloisKeys (from linear transforms)
	// OR in liveRotKeys (from previous AddRotationKey calls)
	keyExists := false
	if scheme.EvalKeys != nil && scheme.EvalKeys.GaloisKeys != nil {
		if _, exists := scheme.EvalKeys.GaloisKeys[galEl]; exists {
			keyExists = true
		}
	}
	if _, exists := liveRotKeys[galEl]; exists {
		keyExists = true
	}

	// Generate the required rotation key if it doesn't exist anywhere
	if !keyExists {
		rotKey := scheme.KeyGen.GenGaloisKeyNew(galEl, scheme.SecretKey)
		liveRotKeys[galEl] = rotKey

		// CRITICAL: Update both scheme.EvalKeys and scheme.Evaluator
		// scheme.EvalKeys is used by EvaluateLinearTransform to create new evaluators
		// Preserve existing keys from scheme.EvalKeys.GaloisKeys (e.g., from linear transforms)
		allKeys := make(map[uint64]*rlwe.GaloisKey)

		// First copy existing keys from scheme.EvalKeys
		if scheme.EvalKeys != nil && scheme.EvalKeys.GaloisKeys != nil {
			for k, v := range scheme.EvalKeys.GaloisKeys {
				allKeys[k] = v
			}
		}

		// Then add/override with keys from liveRotKeys
		for k, v := range liveRotKeys {
			allKeys[k] = v
		}

		// Convert to list for NewMemEvaluationKeySet
		allKeysList := make([]*rlwe.GaloisKey, 0, len(allKeys))
		for _, v := range allKeys {
			allKeysList = append(allKeysList, v)
		}

		scheme.EvalKeys = rlwe.NewMemEvaluationKeySet(scheme.RelinKey, allKeysList...)
		scheme.Evaluator = scheme.Evaluator.WithKey(scheme.EvalKeys)
	}
}

//export Negate
func Negate(ciphertextID C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	ctOut, err := scheme.Evaluator.MulNew(ctIn, -1.0)
	if err != nil {
		panic(err)
	}

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export Rotate
func Rotate(ciphertextID, amount C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	AddRotationKey(amount)
	scheme.Evaluator.Rotate(ctIn, int(amount), ctIn)

	return ciphertextID
}

//export RotateNew
func RotateNew(ciphertextID, amount C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	AddRotationKey(amount)

	ctOut, err := scheme.Evaluator.RotateNew(ctIn, int(amount))
	if err != nil {
		panic(err)
	}

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export Rescale
func Rescale(ciphertextID C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	scheme.Evaluator.Rescale(ctIn, ctIn)

	return ciphertextID
}

//export DropLevel
func DropLevel(ciphertextID C.int, levels C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	scheme.Evaluator.DropLevel(ctIn, int(levels))

	return ciphertextID
}

//export DropLevelNew
func DropLevelNew(ciphertextID C.int, levels C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	ctOut := scheme.Evaluator.DropLevelNew(ctIn, int(levels))

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export SetScale
func SetScale(ciphertextID C.int, scale C.double) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	newScale := rlwe.NewScale(scale)
	scheme.Evaluator.SetScale(ctIn, newScale)

	return ciphertextID
}

//export MatchScalesInPlace
func MatchScalesInPlace(ct0ID, ct1ID C.int) {
	ct0 := RetrieveCiphertext(int(ct0ID))
	ct1 := RetrieveCiphertext(int(ct1ID))

	// Get scales
	scale0 := ct0.Scale.Float64()
	scale1 := ct1.Scale.Float64()

	// Set both to the minimum scale (no rescaling, just adjust scale value)
	// This is similar to SEAL's mod_switch approach
	minScale := scale0
	if scale1 < minScale {
		minScale = scale1
	}

	scheme.Evaluator.SetScale(ct0, rlwe.NewScale(minScale))
	scheme.Evaluator.SetScale(ct1, rlwe.NewScale(minScale))
}

//export ModSwitchTo
func ModSwitchTo(ctID, targetCtID C.int) C.int {
	ct := RetrieveCiphertext(int(ctID))
	targetCt := RetrieveCiphertext(int(targetCtID))

	// Get current and target levels
	currentLevel := ct.Level()
	targetLevel := targetCt.Level()

	// If already at target level, nothing to do
	if currentLevel == targetLevel {
		return ctID
	}

	// If current level is higher, drop to target level (mod switch)
	// This matches SEAL's mod_switch_to_inplace: only drop moduli, DON'T change scale
	if currentLevel > targetLevel {
		levelsToDrop := currentLevel - targetLevel
		scheme.Evaluator.DropLevel(ct, levelsToDrop)
		// NOTE: DropLevel changes only the level, not the scale
		// The scale remains unchanged, which is correct
		return ctID
	}

	// If current level is lower than target, this is an error
	panic("ModSwitchTo: cannot switch to higher level")
}

//export ModSwitchToNew
func ModSwitchToNew(ctID, targetCtID C.int) C.int {
	ct := RetrieveCiphertext(int(ctID))
	targetCt := RetrieveCiphertext(int(targetCtID))

	// Create a copy
	ctCopy := ct.CopyNew()

	// Get current and target levels
	currentLevel := ctCopy.Level()
	targetLevel := targetCt.Level()

	// If already at target level, nothing to do
	if currentLevel == targetLevel {
		idx := PushCiphertext(ctCopy)
		return C.int(idx)
	}

	// If current level is higher, drop to target level (mod switch)
	// This matches SEAL's mod_switch_to_inplace: only drop moduli, DON'T change scale
	if currentLevel > targetLevel {
		levelsToDrop := currentLevel - targetLevel
		ctOut := scheme.Evaluator.DropLevelNew(ctCopy, levelsToDrop)
		// NOTE: DropLevel changes only the level, not the scale
		// The scale remains unchanged, which is correct
		idx := PushCiphertext(ctOut)
		return C.int(idx)
	}

	// If current level is lower than target, this is an error
	panic("ModSwitchToNew: cannot switch to higher level")
}

//export RescaleNew
func RescaleNew(ciphertextID C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	scheme.Evaluator.Rescale(ctIn, ctIn)
	ctOut := ctIn.CopyNew()

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export AddScalar
func AddScalar(ciphertextID C.int, scalar C.float) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	scheme.Evaluator.Add(ctIn, float64(scalar), ctIn)

	return ciphertextID
}

//export AddScalarNew
func AddScalarNew(ciphertextID C.int, scalar C.float) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	ctOut, err := scheme.Evaluator.AddNew(ctIn, float64(scalar))
	if err != nil {
		panic(err)
	}

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export SubScalar
func SubScalar(ciphertextID C.int, scalar C.float) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	scheme.Evaluator.Sub(ctIn, float64(scalar), ctIn)

	return ciphertextID
}

//export SubScalarNew
func SubScalarNew(ciphertextID C.int, scalar C.float) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	ctOut, err := scheme.Evaluator.SubNew(ctIn, float64(scalar))
	if err != nil {
		panic(err)
	}

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export MulScalarInt
func MulScalarInt(ciphertextID C.int, scalar C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	scheme.Evaluator.Mul(ctIn, int(scalar), ctIn)

	return ciphertextID
}

//export MulScalarIntNew
func MulScalarIntNew(ciphertextID C.int, scalar C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	ctOut, err := scheme.Evaluator.MulNew(ctIn, int(scalar))
	if err != nil {
		panic(err)
	}

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export MulScalarFloat
func MulScalarFloat(ciphertextID C.int, scalar C.float) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	scheme.Evaluator.Mul(ctIn, float64(scalar), ctIn)

	return ciphertextID
}

//export MulScalarFloatNew
func MulScalarFloatNew(ciphertextID C.int, scalar C.float) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	ctOut, err := scheme.Evaluator.MulNew(ctIn, float64(scalar))
	if err != nil {
		panic(err)
	}

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export AddPlaintext
func AddPlaintext(ciphertextID, plaintextID C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	ptIn := RetrievePlaintext(int(plaintextID))
	scheme.Evaluator.Add(ctIn, ptIn, ctIn)

	return ciphertextID
}

//export AddPlaintextNew
func AddPlaintextNew(ciphertextID, plaintextID C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	ptIn := RetrievePlaintext(int(plaintextID))

	ctOut, err := scheme.Evaluator.AddNew(ctIn, ptIn)
	if err != nil {
		panic(err)
	}

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export SubPlaintext
func SubPlaintext(ciphertextID, plaintextID C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	ptIn := RetrievePlaintext(int(plaintextID))
	scheme.Evaluator.Sub(ctIn, ptIn, ctIn)

	return ciphertextID
}

//export SubPlaintextNew
func SubPlaintextNew(ciphertextID, plaintextID C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	ptIn := RetrievePlaintext(int(plaintextID))

	ctOut, err := scheme.Evaluator.SubNew(ctIn, ptIn)
	if err != nil {
		panic(err)
	}

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export MulPlaintext
func MulPlaintext(ciphertextID, plaintextID C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	ptIn := RetrievePlaintext(int(plaintextID))
	scheme.Evaluator.Mul(ctIn, ptIn, ctIn)

	return ciphertextID
}

//export MulPlaintextNew
func MulPlaintextNew(ciphertextID, plaintextID C.int) C.int {
	ctIn := RetrieveCiphertext(int(ciphertextID))
	ptIn := RetrievePlaintext(int(plaintextID))

	ctOut, err := scheme.Evaluator.MulNew(ctIn, ptIn)
	if err != nil {
		panic(err)
	}

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export AddCiphertext
func AddCiphertext(ctID0, ctID1 C.int) C.int {
	ctIn0 := RetrieveCiphertext(int(ctID0))
	ctIn1 := RetrieveCiphertext((int(ctID1)))
	scheme.Evaluator.Add(ctIn0, ctIn1, ctIn0)

	return ctID0
}

//export AddCiphertextNew
func AddCiphertextNew(ctID0, ctID1 C.int) C.int {
	ctIn0 := RetrieveCiphertext(int(ctID0))
	ctIn1 := RetrieveCiphertext((int(ctID1)))

	ctOut, err := scheme.Evaluator.AddNew(ctIn0, ctIn1)
	if err != nil {
		panic(err)
	}

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export SubCiphertext
func SubCiphertext(ctID0, ctID1 C.int) C.int {
	ctIn0 := RetrieveCiphertext(int(ctID0))
	ctIn1 := RetrieveCiphertext((int(ctID1)))
	scheme.Evaluator.Sub(ctIn0, ctIn1, ctIn0)

	return ctID0
}

//export SubCiphertextNew
func SubCiphertextNew(ctID0, ctID1 C.int) C.int {
	ctIn0 := RetrieveCiphertext(int(ctID0))
	ctIn1 := RetrieveCiphertext((int(ctID1)))

	ctOut, err := scheme.Evaluator.SubNew(ctIn0, ctIn1)
	if err != nil {
		panic(err)
	}

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

//export MulRelinCiphertext
func MulRelinCiphertext(ctID0, ctID1 C.int) C.int {
	ctIn0 := RetrieveCiphertext(int(ctID0))
	ctIn1 := RetrieveCiphertext((int(ctID1)))
	scheme.Evaluator.MulRelin(ctIn0, ctIn1, ctIn0)

	return ctID0
}

//export MulRelinCiphertextNew
func MulRelinCiphertextNew(ctID0, ctID1 C.int) C.int {
	ctIn0 := RetrieveCiphertext(int(ctID0))
	ctIn1 := RetrieveCiphertext((int(ctID1)))

	ctOut, err := scheme.Evaluator.MulRelinNew(ctIn0, ctIn1)
	if err != nil {
		panic(err)
	}

	idx := PushCiphertext(ctOut)
	return C.int(idx)
}

func DeleteRotationKeys() {
	liveRotKeys = make(map[uint64]*rlwe.GaloisKey)
	savedRotKeys = []uint64{}
}

File: orion/backend/python/evaluator.py
import h5py


class NewEvaluator:
    def __init__(self, scheme):
        self.backend = scheme.backend
        self.params = scheme.params
        self.io_mode = self.params.get_io_mode()
        self.keys_path = self.params.get_keys_path()
        self.loaded_rotation_keys = set()
        self.new_evaluator()
        if self.io_mode in ("none", "save"):
            self._initialize_power_of_two_rotation_keys()

    def new_evaluator(self):
        self.backend.NewEvaluator()

    def add_rotation_key(self, amount: int):
        self.backend.AddRotationKey(amount)

    def _power_of_two_rotations(self):
        amount = 1
        while amount <= self.params.get_slots():
            yield amount
            amount *= 2

    def _galois_element(self, amount):
        return int(self.backend.GetRotationGaloisElement(amount))

    def _initialize_power_of_two_rotation_keys(self):
        if self.io_mode == "none":
            for amount in self._power_of_two_rotations():
                self.add_rotation_key(amount)
            return

        with h5py.File(self.keys_path, "a") as keys:
            for amount in self._power_of_two_rotations():
                galois_element = self._galois_element(amount)
                key_name = str(galois_element)
                if key_name in keys:
                    continue
                serialized, pointer = self.backend.GenerateAndSerializeRotationKey(
                    galois_element
                )
                try:
                    keys.create_dataset(key_name, data=serialized)
                finally:
                    self.backend.FreeCArray(pointer)

    def _load_rotation_key(self, amount):
        galois_element = self._galois_element(amount)
        if galois_element in self.loaded_rotation_keys:
            return
        with h5py.File(self.keys_path, "r") as keys:
            serialized = keys[str(galois_element)][()]
        self.backend.LoadRotationKey(serialized, galois_element)
        self.loaded_rotation_keys.add(galois_element)

    def preload_power_of_two_rotation_keys(self):
        if self.io_mode == "load":
            for amount in self._power_of_two_rotations():
                self._load_rotation_key(amount)

    def negate(self, ctxt):
        return self.backend.Negate(ctxt)
    
    def rotate(self, ctxt, amount, in_place):
        if self.io_mode == "load":
            self._load_rotation_key(amount)
        if in_place:
            return self.backend.Rotate(ctxt, amount)
        return self.backend.RotateNew(ctxt, amount)

    def add_scalar(self, ctxt, scalar, in_place):
        if in_place:
            return self.backend.AddScalar(ctxt, float(scalar))
        return self.backend.AddScalarNew(ctxt, float(scalar))

    def sub_scalar(self, ctxt, scalar, in_place):
        if in_place:
            return self.backend.SubScalar(ctxt, float(scalar))
        return self.backend.SubScalarNew(ctxt, float(scalar))

    def mul_scalar(self, ctxt, scalar, in_place):
        if isinstance(scalar, float) and scalar.is_integer():
            scalar = int(scalar)  # (e.g., 1.00 -> 1)

        if isinstance(scalar, int):
            ct_out = (self.backend.MulScalarInt if in_place 
                      else self.backend.MulScalarIntNew)(ctxt, scalar)
        else:
            ct_out = (self.backend.MulScalarFloat if in_place 
                      else self.backend.MulScalarFloatNew)(ctxt, scalar)
            ct_out = self.backend.Rescale(ct_out)

        return ct_out
        
    def add_plaintext(self, ctxt, ptxt, in_place):
        if in_place:
            return self.backend.AddPlaintext(ctxt, ptxt) 
        return self.backend.AddPlaintextNew(ctxt, ptxt) 

    def sub_plaintext(self, ctxt, ptxt, in_place):
        if in_place:
            return self.backend.SubPlaintext(ctxt, ptxt) 
        return self.backend.SubPlaintextNew(ctxt, ptxt) 

    def mul_plaintext(self, ctxt, ptxt, in_place):
        if in_place: # ct_out = ctxt
            ct_out = self.backend.MulPlaintext(ctxt, ptxt)
        else:
            ct_out = self.backend.MulPlaintextNew(ctxt, ptxt) 
        
        return self.backend.Rescale(ct_out)

    def add_ciphertext(self, ctxt0, ctxt1, in_place):
        if in_place:
            return self.backend.AddCiphertext(ctxt0, ctxt1)
        return self.backend.AddCiphertextNew(ctxt0, ctxt1)

    def sub_ciphertext(self, ctxt0, ctxt1, in_place):
        if in_place:
            return self.backend.SubCiphertext(ctxt0, ctxt1)
        return self.backend.SubCiphertextNew(ctxt0, ctxt1)

    def mul_ciphertext(self, ctxt0, ctxt1, in_place):
        if in_place: # ct_out = ctxt
            ct_out = self.backend.MulRelinCiphertext(ctxt0, ctxt1)
        else:
            ct_out = self.backend.MulRelinCiphertextNew(ctxt0, ctxt1)
        
        return self.backend.Rescale(ct_out)
    
    def rescale(self, ctxt, in_place):
        if in_place:
            return self.backend.Rescale(ctxt)
        return self.backend.RescaleNew(ctxt)
    
    def get_live_plaintexts(self):
        return self.backend.GetLivePlaintexts() 

    def get_live_ciphertexts(self):
        return self.backend.GetLiveCiphertexts()
